File: leetcode-submissions/64_Minimum_Path_Sum.cpp
class Solution {
public:
vector<vector<int>>arr;
int n,m;
int f(int i,int j){
    

    //base cases
    if(i>=n or j>=m or i<0 or j<0 ){
        return INT_MAX;
    } 
    if(i==n-1 and j==m-1){
        return arr[n-1][m-1];
    }
    return arr[i][j]+min(f(i+1,j),f(i,j+1));
}
    int minPathSum(vector<vector<int>>& grid) {
    arr=grid;
     n=arr.size();
     m=arr[0].size();
    return f(0,0);
    }
};

File: leetcode-submissions/64_Minimum_Path_Sum.ts
function minPathSum(grid: number[][]): number {

    const m  = grid.length , n = grid[0].length;
    const dp : Array<Array<number>> = Array.from({length:m+1}, ()=>Array.from({length:n+1}, ()=>-1));
    return solve( 0 , 0 , m , n , dp , grid);

};
function solve( i:number , j:number , m :number , n :number , dp:Array<Array<number>>, grid:Array<Array<number>>){
    if(i===m-1 && j===n-1){
        return  grid[i][j] ;

    }
    if(i<0 || j<0 || j>=n || i>=m ){
        return Number.MAX_SAFE_INTEGER;

    }
    if(dp[i][j]!==-1){
        return dp[i][j];
    }
    const downSum :number = grid[i][j]+solve(i+1 , j , m , n , dp , grid);
    const rightSum :number =  grid[i][j]+solve(i , j+1 , m , n , dp , grid);
    return dp[i][j]=Math.min(downSum , rightSum);
}
